server/src/main/java/org/elasticsearch/action/support/tasks/TransportTasksAction.java

```diff
@@ -85,66 +81,164 @@ protected TransportTasksAction(
 
     @Override
     protected void doExecute(Task task, TasksRequest request, ActionListener<TasksResponse> listener) {
-        new AsyncAction(task, request, listener).start();
-    }
 
-    private void nodeOperation(CancellableTask task, NodeTaskRequest nodeTaskRequest, ActionListener<NodeTasksResponse> listener) {
-        TasksRequest request = nodeTaskRequest.tasksRequest;
-        processTasks(request, ActionListener.wrap(tasks -> nodeOperation(task, listener, request, tasks), listener::onFailure));
+        final var taskResponses = new ArrayList<TaskResponse>();
+        final var taskOperationFailures = new ArrayList<TaskOperationFailure>();
+        final var failedNodeExceptions = new ArrayList<FailedNodeException>();
+
+        final var resultListener = new SubscribableListener<TasksResponse>();
+        final var resultListenerCompleter = new RunOnce(() -> {
+            if (task instanceof CancellableTask cancellableTask && cancellableTask.notifyIfCancelled(resultListener)) {
+                return;
+            }
+            // ref releases all happen-before here so no need to be synchronized
+            ActionListener.completeWith(
+                resultListener,
+                () -> newResponse(request, taskResponses, taskOperationFailures, failedNodeExceptions)
+            );
+        });
+
+        // collects node listeners & completes them if cancelled
+        final var nodeCancellationListener = new SubscribableListener<NodeTasksResponse>();
+        if (task instanceof CancellableTask cancellableTask) {
+            cancellableTask.addListener(() -> {
+                assert cancellableTask.isCancelled();
+                resultListenerCompleter.run();
+                cancellableTask.notifyIfCancelled(nodeCancellationListener);
+            });
+        }
+
+        try (var refs = new RefCountingRunnable(() -> {
+            resultListener.addListener(listener);
+            resultListenerCompleter.run();
+        })) {
+            final var discoveryNodes = clusterService.state().nodes();
+            final String[] nodeIds = resolveNodes(request, discoveryNodes);
+
+            final var transportRequestOptions = TransportRequestOptions.timeout(request.getTimeout());
+
+            for (final var nodeId : nodeIds) {
+                final ActionListener<NodeTasksResponse> nodeResponseListener = ActionListener.notifyOnce(new ActionListener<>() {
+                    @Override
+                    public void onResponse(NodeTasksResponse nodeResponse) {
+                        synchronized (taskResponses) {
+                            taskResponses.addAll(nodeResponse.results);
```

Comment:
I've run some tests and have seen a lot of cases where `nodeResponse.results` is empty. Would it make sense to perform an `isEmpty` check on `nodeResponse.results` before acquiring the lock on `taskResponses`?